turtlebot3_gazebo/src/turtlebot3_drive.cpp

```diff
@@ -0,0 +1,179 @@
+/*******************************************************************************
+* Copyright 2019 ROBOTIS CO., LTD.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/* Authors: Taehun Lim (Darby), Ryan Shim */
+
+#include "turtlebot3_gazebo/turtlebot3_drive.hpp"
+
+using namespace std::chrono_literals;
+
+Turtlebot3Drive::Turtlebot3Drive()
+: Node("turtlebot3_drive_node")
+{
+  /************************************************************
+  ** Initialise variables
+  ************************************************************/
+  scan_data_[0] = 0.0;
+  scan_data_[1] = 0.0;
+  scan_data_[2] = 0.0;
+
+  robot_pose_ = 0.0;
+  prev_robot_pose_ = 0.0;
+
+  /************************************************************
+  ** Initialise ROS publishers and subscribers
+  ************************************************************/
+  auto qos = rclcpp::QoS(rclcpp::KeepLast(10));
+
+  // Initialise publishers
+  cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", qos);
+
+  // Initialise subscribers
+  scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
+    "scan", rclcpp::SensorDataQoS(), std::bind(&Turtlebot3Drive::scan_callback, this, std::placeholders::_1));
+  odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
+    "odom", qos, std::bind(&Turtlebot3Drive::odom_callback, this, std::placeholders::_1));
+
+  /************************************************************
+  ** Initialise ROS timers
+  ************************************************************/
+  update_timer_ = this->create_wall_timer(10ms, std::bind(&Turtlebot3Drive::update_callback, this));
+
+  RCLCPP_INFO(this->get_logger(), "Turtlebot3 simulation node has been initialised");
+}
+
+Turtlebot3Drive::~Turtlebot3Drive()
+{
+  RCLCPP_INFO(this->get_logger(), "Turtlebot3 simulation node has been terminated");
+}
+
+/********************************************************************************
+** Callback functions for ROS subscribers
+********************************************************************************/
+void Turtlebot3Drive::odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
+{
+  tf2::Quaternion q(
+        msg->pose.pose.orientation.x,
+        msg->pose.pose.orientation.y,
+        msg->pose.pose.orientation.z,
+        msg->pose.pose.orientation.w);
```

Comment:
It seems like you used too much indentation.

Reply:
changed